ludwig/encoders/text_encoders.py

```diff
@@ -2374,3 +2380,150 @@ def output_shape(self) -> torch.Size:
 
     def get_embedding_layer(self) -> nn.Module:
         return self
+
+
+@DeveloperAPI
+@register_encoder("llm", [TEXT])
+class LLMEncoder(Encoder):
+    # Per-adapter type prefixes for parameter names in the state dict, taken from
+    # https://github.com/huggingface/peft/blob/0f1e9091cc975eb5458cc163bf1843a34fb42b76/src/peft/utils/save_and_load.py#L173C9-L180
+    ADAPTER_PARAM_NAME_PREFIX = {"lora": "lora_", "adalora": "lora_"}
+
+    def __init__(self, encoder_config: LLMEncoderConfig = None, **kwargs):
+        super().__init__()
+        self.config = encoder_config
+
+        self.model_name = self.config.base_model
+        self.model_config = AutoConfig.from_pretrained(self.config.base_model)
+
+        self.model = load_pretrained_from_config(self.config, model_config=self.model_config)
+        self.curr_device = next(self.model.parameters()).device
+        logger.info("Done.")
+
+        self.context_len = get_context_len(self.model_config)
+
+        # TODO(Arnav): This needs be more flexible to account for RoPE Scaling
+        # When merging input IDs and target IDs for LLM fine-tuning, we want to make sure that the merged tensor is
+        # not longer than the global maximum sequence length. This is provided in the preprocessing config. We never
+        # want to exceed the maximum possible context length so we also check for that.
+        if self.config.max_sequence_length:
+            max_sequence_length = self.config.max_sequence_length
+            self.max_sequence_length = (
+                max_sequence_length if max_sequence_length <= self.context_len else self.context_len
+            )
+        else:
+            self.max_sequence_length = self.context_len
+
+        # Initialize tokenizer
+        self.tokenizer = HFTokenizer(self.config.base_model).tokenizer
+
+        self.attention_masks = None
+
+        self.prepare_for_training()
+
+        if not self.config.adapter:
+            out_module = list(self.model.modules())[-1]
+            out_module.requires_grad_(requires_grad=False)
```

Comment:
Can we add a comment explaining what this code does, i.e., why are we turning off gradients for the last module? I'd also recommend moving this to within the `prepare_for_training()` encapsulation since it seems in spirit of that function!

Reply:
Done!